sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1648,16 +1648,56 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
   }
 
   /**
+   * Statically evaluate an expression containing one or more aggregates on an empty input.
+   */
+  private def evalOnZeroTups(expr : Expression) : Option[Any] = {
+    // AggregateExpressions are Unevaluable, so we need to replace all aggregates
+    // in the expression with the value they would return for zero input tuples.
+    val rewrittenExpr = expr transform {
+      case a @ AggregateExpression(aggFunc, _, _, resultId) =>
+        val resultLit = aggFunc.defaultResult match {
+          case Some(lit) => lit
+          case None => Literal.default(NullType)
+        }
+        Alias(resultLit, "aggVal") (exprId = resultId)
+    }
+    Option(rewrittenExpr.eval())
+  }
+
+  /**
    * Construct a new child plan by left joining the given subqueries to a base plan.
    */
   private def constructLeftJoins(
       child: LogicalPlan,
       subqueries: ArrayBuffer[ScalarSubquery]): LogicalPlan = {
     subqueries.foldLeft(child) {
       case (currentChild, ScalarSubquery(query, conditions, _)) =>
+        val aggOutputExpr = query.asInstanceOf[Aggregate].aggregateExpressions.head
+        val origOutput = query.output.head
+
+        // Ensure the rewritten subquery returns the same result when a tuple from the
+        // outer query block does not join with the subquery block.
+        // val (outputExpr, rewrittenQuery) = aggFunc.defaultResult match {
+        val (outputExpr, rewrittenQuery) = evalOnZeroTups(aggOutputExpr) match {
+          case Some(value) =>
+            val origExprId = origOutput.exprId
+            val newExprId = NamedExpression.newExprId
+
+            // Renumber the original output, because the outer query refers to its ID.
+            val newQuery = query transformExpressions {
+              case Alias(c, n) => Alias(c, n)(exprId = newExprId)
```

Comment:
Oops, there should be an "if" after that "case" expression. I'll update that expression to only renumber `origExprId` to `newExprId`.